File: typescript/src/__tests__/query.test.ts
import { describe, it, expect } from "vitest";
import { Engine, ExpressionEnv } from "../index.js";

function createEngine() {
  return new Engine(new ExpressionEnv());
}

describe("query expressions", () => {
  const engine = createEngine();

  it("basic $expr $pattern query", () => {
    const query = {
      $expr: ["$pattern", "$*", "author of", "$*"],
    };

    const result = engine.execute(query) as string;

    expect(result).toContain("select");
    expect(result).toContain("subject, predicate, object, meta");
    expect(result).toContain("from statement as s");
    expect(result).toContain("author of");
    expect(result).toContain("triple");
    expect(result).toContain("offset 0");
    expect(result).toContain("limit 100");
  });

  it("combined $query with $and patterns", () => {
    const query = {
      $query: [
        "$and",
        [
          ["$pattern", "Liu Xin", "author of", "$*"],
          ["$pattern", "$*", "author of", "$*"],
        ],
      ],
    };

    const result = engine.execute(query) as string;

    expect(result).toContain("select subject, predicate, object, meta");
    expect(result).toContain("from statement");
    expect(result).toContain("Liu Xin");
    expect(result).toContain("author of");
    expect(result).toContain(" and ");
    expect(result).toContain("offset 0");
    expect(result).toContain("limit 100");
  });
});

File: java/src/test/java/io/github/marchliu/jse/QueryTest.java
package io.github.marchliu.jse;

import io.github.marchliu.jse.functors.SqlFunctors;
import io.github.marchliu.jse.functors.UtilsFunctors;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertTrue;

class QueryTest {

    private Engine createEngine() {
        Env env = new Env();
        env.load(UtilsFunctors.UTILS_FUNCTORS);
        env.load(SqlFunctors.SQL_FUNCTORS);
        return new Engine(env);
    }

    @Test
    void basicQuery() {
        Engine engine = createEngine();
        Map<String, Object> query = Map.of(
                "$query", List.of("$quote", List.of("$pattern", "$*", "author of", "$*"))
        );

        Object result = engine.execute(query);
        String sql = (String) result;

        // Loose checking - look for keywords, not exact format
        assertTrue(sql.contains("select"));
        assertTrue(sql.contains("subject, predicate, object, meta"));
        assertTrue(sql.contains("from statement"));
        assertTrue(sql.contains("author of"));
        assertTrue(sql.contains("triple"));
        assertTrue(sql.contains("offset 0"));
        assertTrue(sql.contains("limit 100"));
    }

    @Test
    void combinedQuery() {
        Engine engine = createEngine();
        Map<String, Object> query = Map.of(
                "$query", Map.of(
                        "$quote", List.of(
                                "$and",
                                List.of("$pattern", "Liu Xin", "author of", "$*"),
                                List.of("$pattern", "$*", "author of", "$*")
                        )
                )
        );

        Object result = engine.execute(query);
        String sql = (String) result;

        assertTrue(sql.contains("select"));
        assertTrue(sql.contains("subject, predicate, object, meta"));
        assertTrue(sql.contains("from statement"));
        assertTrue(sql.contains("Liu Xin"));
        assertTrue(sql.contains("author of"));
        assertTrue(sql.contains(" and "));
        assertTrue(sql.contains("offset 0"));
        assertTrue(sql.contains("limit 100"));
    }
}
